web-client/src/views/FileHandlingHelpers/fileValidation.ts

```diff
@@ -0,0 +1,204 @@
+  }
+  return `The file is not in a supported format (${fileExtensions.join(', ')}). Select a different file or resave it in a supported format.`;
+};
+
+export const validateFileSize = ({
+  file,
+  megabyteLimit,
+}: {
+  file: File;
+  megabyteLimit: number;
+}): FileValidationResponse => {
+  if (file.size > megabyteLimit * 1024 * 1024) {
+    return {
+      errorInformation: {
+        errorMessageToDisplay: `The file size is too big. The maximum file size is ${megabyteLimit}MB. Reduce the file size and try again.`,
+        errorType: ErrorTypes.FILE_TOO_BIG,
+      },
+      isValid: false,
+    };
+  }
+  return { isValid: true };
+};
+
+const validateCorrectFileType = ({
+  allowedFileExtensions,
+  file,
+}: {
+  file: File;
+  allowedFileExtensions: string[];
+}): FileValidationResponse => {
+  if (!allowedFileExtensions.includes(getFileExtension(file.name))) {
+    return {
+      errorInformation: {
+        errorMessageToDisplay: getWrongFileTypeMessage(allowedFileExtensions),
+        errorType: ErrorTypes.WRONG_FILE_TYPE,
+      },
+      isValid: false,
+    };
+  }
+  return { isValid: true };
+};
+
+export const validateFileOnSelect = async ({
+  allowedFileExtensions,
+  e,
+  megabyteLimit,
+  onError,
+  onSuccess,
+}: {
+  allowedFileExtensions: string[];
+  e: React.ChangeEvent<HTMLInputElement>;
+  megabyteLimit: number;
+  onSuccess: ({ selectedFile }: { selectedFile: File }) => void;
+  onError: ({
+    errorType,
+    messageToDisplay,
+    messageToLog,
+  }: {
+    messageToDisplay: string;
+    errorType?: ErrorTypes;
+    messageToLog?: string;
+  }) => void;
+}) => {
+  const target = e.target as HTMLInputElement;
+
+  if (!target || !target.files || target.files.length === 0) {
+    onError({
+      messageToDisplay: 'No file selected. Please upload a file.',
+    });
+    return;
+  }
+
+  const selectedFile = target.files[0];
+  const { errorInformation, isValid } = await validateFile({
+    allowedFileExtensions,
+    file: selectedFile,
+    megabyteLimit,
+  });
+  if (!isValid) {
+    onError({
+      errorType: errorInformation?.errorType,
+      messageToDisplay: errorInformation?.errorMessageToDisplay!,
+      messageToLog: errorInformation?.errorMessageToLog,
+    });
+    target.value = '';
+    return;
+  }
+  onSuccess({ selectedFile });
+};
+
+export const validateFile = async ({
+  allowedFileExtensions,
+  file,
+  megabyteLimit,
+}: {
+  file: File;
+  megabyteLimit: number;
+  allowedFileExtensions: string[];
+}): Promise<FileValidationResponse> => {
+  const fileSizeValidation = validateFileSize({ file, megabyteLimit });
+  if (!fileSizeValidation.isValid) {
+    return fileSizeValidation;
+  }
+  const correctFileValidation = validateCorrectFileType({
+    allowedFileExtensions,
+    file,
+  });
+  if (!correctFileValidation.isValid) {
+    return correctFileValidation;
+  }
+  if (file.type === 'application/pdf') {
+    return await validatePdf({ file });
```

Comment:
I have a couple of questions on this piece -

* Can we trust this to always be populated by the built-in nodejs `fs` package? (false positives)
* Are there any legitimate PDF files (hi-res, interactive, etc) that don't get this type? (false negatives)

Reply:
Good question!

In this case, I don't think node's `fs` is involved. The file type is being determined by the browser, usually via the `File` API, which typically uses the file extension but may also infer the MIME type from the file's header. This detection _isn't_ always perfect. A legitimate PDF saved with a wrong extension (e.g., .csv) will likely not be recognized as application/pdf. Conversely, a file mislabeled as a PDF (e.g., a .csv file renamed to .pdf) might pass the MIME type check, but then the PDF validation logic would catch the mismatch and flag it as invalid.

Regarding different types of PDFs (such as PDF/A or PDF/E), our current validation logic doesn't support them, so files with those specific formats may fail validation, even though they are legitimate PDFs. (But this is already the case!)

Does that make sense and answer your questions?

Reply:
Yes it does, thanks!